flink/src/main/java/org/apache/iceberg/flink/source/StreamingReaderOperator.java

```diff
@@ -0,0 +1,216 @@
+    this.processingTimeService = timeService;
+    this.executor = (MailboxExecutorImpl) Preconditions.checkNotNull(mailboxExecutor,
+        "The mailboxExecutor should not be null.");
+  }
+
+  @Override
+  public void initializeState(StateInitializationContext context) throws Exception {
+    super.initializeState(context);
+
+    checkpointState = context.getOperatorStateStore().getListState(
+        new ListStateDescriptor<>("splits", new JavaSerializer<>()));
+
+    int subtaskIdx = getRuntimeContext().getIndexOfThisSubtask();
+    splits = Lists.newLinkedList();
+    if (context.isRestored()) {
+      LOG.info("Restoring state for the {} (taskIdx={}).", getClass().getSimpleName(), subtaskIdx);
+
+      for (FlinkInputSplit split : checkpointState.get()) {
+        splits.add(split);
+      }
+    }
+
+    this.readerContext = StreamSourceContexts.getSourceContext(
+        getOperatorConfig().getTimeCharacteristic(),
+        getProcessingTimeService(),
+        new Object(), // no actual locking needed
+        getContainingTask().getStreamStatusMaintainer(),
+        output,
+        getRuntimeContext().getExecutionConfig().getAutoWatermarkInterval(),
+        -1);
+
+    if (!splits.isEmpty()) {
+      enqueueProcessSplits();
+    }
+  }
+
+  @Override
+  public void processElement(StreamRecord<FlinkInputSplit> element) {
+    splits.offer(element.getValue());
+    enqueueProcessSplits();
+  }
+
+  private void enqueueProcessSplits() {
+    executor.execute(this::processSplits, this.getClass().getSimpleName());
+  }
+
+  private void processSplits() throws IOException {
+    do {
+      FlinkInputSplit split = splits.poll();
+      if (split == null) {
+        return;
+      }
+
+      LOG.debug("load split: {}", split);
+      format.open(split);
+
+      RowData nextElement = null;
+      while (!format.reachedEnd()) {
+        nextElement = format.nextRecord(nextElement);
+        if (nextElement != null) {
+          readerContext.collect(nextElement);
+        } else {
+          break;
+        }
+      }
+
+      format.close();
+    } while (executor.isIdle());
+    enqueueProcessSplits();
+  }
+
+  @Override
+  public void processWatermark(Watermark mark) {
+    // we do nothing because we emit our own watermarks if needed.
+  }
+
+  @Override
+  public void dispose() throws Exception {
+    super.dispose();
+
+    if (format != null) {
+      format.close();
+      format.closeInputFormat();
+      format = null;
+    }
+
+    readerContext = null;
+  }
+
+  @Override
+  public void close() throws Exception {
+    super.close();
+    output.close();
+    if (readerContext != null) {
+      readerContext.emitWatermark(Watermark.MAX_WATERMARK);
+      readerContext.close();
+      readerContext = null;
+    }
+  }
+
+  @Override
+  public void snapshotState(StateSnapshotContext context) throws Exception {
+    super.snapshotState(context);
+
+    checkpointState.clear();
+    for (FlinkInputSplit split : splits) {
```

Comment:
should we use `addAll` instead, which translates to one `merge` call in rocksdb?